lighthouse-core/computed/metrics/lantern-first-contentful-paint.js

```diff
@@ -24,52 +26,152 @@ class LanternFirstContentfulPaint extends LanternMetric {
+    if (firstParse) blockingCpuNodeIds.add(firstParse.id);
+
+    // If a CPU filter was passed in, we also want to include those extra nodes.
+    if (extraBlockingCpuNodesToIncludeFilter) {
+      cpuNodes
+        .filter(extraBlockingCpuNodesToIncludeFilter)
+        .forEach(node => blockingCpuNodeIds.add(node.id));
+    }
+
+    return {
+      definitelyNotRenderBlockingScriptUrls,
+      blockingCpuNodeIds,
+    };
   }
 
   /**
+   * This function computes the graph required for the first paint of interest.
+   *
    * @param {Node} dependencyGraph
-   * @param {LH.Artifacts.TraceOfTab} traceOfTab
+   * @param {number} paintTs
+   * @param {function(NetworkNode):boolean} blockingScriptFilter
+   * @param {function(CPUNode):boolean=} extraBlockingCpuNodesToIncludeFilter
    * @return {Node}
    */
-  static getPessimisticGraph(dependencyGraph, traceOfTab) {
-    const fcp = traceOfTab.timestamps.firstContentfulPaint;
-    const blockingScriptUrls = LanternMetric.getScriptUrls(dependencyGraph, node => {
-      return node.endTime <= fcp && node.hasRenderBlockingPriority();
-    });
+  static getFirstPaintBasedGraph(
+      dependencyGraph,
+      paintTs,
+      blockingScriptFilter,
+      extraBlockingCpuNodesToIncludeFilter
+  ) {
+    const {
+      definitelyNotRenderBlockingScriptUrls,
+      blockingCpuNodeIds,
+    } = this.getBlockingCpuData(
+      dependencyGraph,
+      paintTs,
+      blockingScriptFilter,
+      extraBlockingCpuNodesToIncludeFilter
+    );
 
     return dependencyGraph.cloneWithRelationships(node => {
-      if (node.endTime > fcp && !node.isMainDocument()) return false;
-      // Include EvaluateScript tasks for blocking scripts
-      if (node.type === BaseNode.TYPES.CPU) {
-        return node.isEvaluateScriptFor(blockingScriptUrls);
-      }
+      if (node.type === BaseNode.TYPES.NETWORK) {
+        // Exclude all nodes that ended after FCP (except for the main document which we always consider necessary)
+        if (node.endTime > paintTs && !node.isMainDocument()) return false;
 
-      // Include non-script-initiated network requests with a render-blocking priority
-      return node.hasRenderBlockingPriority();
+        const url = node.record.url;
+        // If the URL definitely wasn't render-blocking then we filter it out.
+        if (definitelyNotRenderBlockingScriptUrls.has(url)) {
+          return false;
+        }
+        return node.hasRenderBlockingPriority();
+      } else {
+        // If it's a CPU node, just check if it was blocking.
+        return blockingCpuNodeIds.has(node.id);
+      }
     });
   }
+
+  /**
+   * @param {Node} dependencyGraph
+   * @param {LH.Artifacts.TraceOfTab} traceOfTab
+   * @return {Node}
+   */
+  static getOptimisticGraph(dependencyGraph, traceOfTab) {
+    return this.getFirstPaintBasedGraph(
+      dependencyGraph,
+      traceOfTab.timestamps.firstContentfulPaint,
+      node => node.hasRenderBlockingPriority() && node.initiatorType !== 'script'
```

Comment:
would a comment (here and in FMP) on the addition of `&& node.initiatorType !== 'script'` for the optimistic graph be good? I don't want to get redundant with other source of information, but not immediately clear while reading this why there'd be this difference between this and the pessimistic graph

Reply:
sure sg 👍